src/pkcs15init/profile.c

```diff
@@ -499,6 +499,8 @@ sc_profile_get_pin_info(struct sc_profile *profile,
 		int id, struct sc_pkcs15_auth_info *info)
 {
 	struct pin_info	*pi;
+	struct sc_pkcs15_auth_info info_tmp = {0};
+	*info = info_tmp;
```

Comment:
I do not think this is correct. In case of the failure, you will pass out of the function the pointer to the local variable, which will no longer exist after you `return`. I think you should just initialize the pointer to NULL.

Reply:
well, this tmp variable is a no-op, just to please gcc compiler.
When I tried with memset, for some reasons it didn't work, and it works now.
pushed.

Reply:
Do you have the specific error message you were getting?

In any case, I think we should first check the pointer is not null before dereferencing it.

Reply:
sorry I did only local builds...

Reply:
are we good? do you want any more log?
```
/bin/bash ../../../libtool  --tag=CC   --mode=link gcc -g -O0   -pthread -Wall -Wextra -Wno-unused-parameter -Werror -Wstrict-aliasing=2 -g -O2 -ffile-prefix-map=/build/opensc-0.23.0=. -flto=auto -ffat-lto-objects -fstack-protector-strong -Wformat -Werror=format-security -fdebug-prefix-map=/build/opensc-0.23.0=/usr/src/opensc-0.23.0-1ubuntu3 -Wno-error=deprecated-declarations -Wno-error=stringop-overflow  -Wl,-Bsymbolic-functions -flto=auto -ffat-lto-objects -Wl,-z,relro -o fuzz_piv_tool fuzz_piv_tool.o fuzzer_reader.o fuzzer_tool.o fuzzer.o ../../tools/util.o -lcrypto   ../../../src/libopensc/libopensc.la ../../../src/common/libscdl.la ../../../src/pkcs15init/libpkcs15init.la ../../../src/common/libcompat.la
/bin/bash ../../../libtool  --tag=CC   --mode=link gcc -g -O0   -pthread -Wall -Wextra -Wno-unused-parameter -Werror -Wstrict-aliasing=2 -g -O2 -ffile-prefix-map=/build/opensc-0.23.0=. -flto=auto -ffat-lto-objects -fstack-protector-strong -Wformat -Werror=format-security -fdebug-prefix-map=/build/opensc-0.23.0=/usr/src/opensc-0.23.0-1ubuntu3 -Wno-error=deprecated-declarations -Wno-error=stringop-overflow -static -Wl,-Bsymbolic-functions -flto=auto -ffat-lto-objects -Wl,-z,relro -o fuzz_pkcs15init fuzz_pkcs15init.o fuzzer_reader.o fuzzer.o -lcrypto  -leac -lcrypto   ../../../src/libopensc/libopensc.la ../../../src/common/libscdl.la ../../../src/pkcs15init/libpkcs15init.la ../../../src/common/libcompat.la
libtool: link: gcc -g -O0 -Wall -Wextra -Wno-unused-parameter -Werror -Wstrict-aliasing=2 -g -O2 -ffile-prefix-map=/build/opensc-0.23.0=. -flto=auto -ffat-lto-objects -fstack-protector-strong -Wformat -Werror=format-security -fdebug-prefix-map=/build/opensc-0.23.0=/usr/src/opensc-0.23.0-1ubuntu3 -Wno-error=deprecated-declarations -Wno-error=stringop-overflow -Wl,-Bsymbolic-functions -flto=auto -ffat-lto-objects -Wl,-z -Wl,relro -o .libs/fuzz_piv_tool fuzz_piv_tool.o fuzzer_reader.o fuzzer_tool.o fuzzer.o ../../tools/util.o  -lcrypto ../../../src/libopensc/.libs/libopensc.so ../../../src/common/.libs/libscdl.a -ldl ../../../src/pkcs15init/.libs/libpkcs15init.a ../../../src/common/.libs/libcompat.a -pthread
libtool: link: gcc -g -O0 -Wall -Wextra -Wno-unused-parameter -Werror -Wstrict-aliasing=2 -g -O2 -ffile-prefix-map=/build/opensc-0.23.0=. -flto=auto -ffat-lto-objects -fstack-protector-strong -Wformat -Werror=format-security -fdebug-prefix-map=/build/opensc-0.23.0=/usr/src/opensc-0.23.0-1ubuntu3 -Wno-error=deprecated-declarations -Wno-error=stringop-overflow -Wl,-Bsymbolic-functions -flto=auto -ffat-lto-objects -Wl,-z -Wl,relro -o fuzz_pkcs15init fuzz_pkcs15init.o fuzzer_reader.o fuzzer.o  ../../../src/libopensc/.libs/libopensc.a -lz -lgio-2.0 -lgobject-2.0 -leac -lcrypto ../../../src/common/.libs/libscdl.a -ldl ../../../src/pkcs15init/.libs/libpkcs15init.a ../../../src/common/.libs/libcompat.a -pthread
../../pkcs15init/pkcs15-entersafe.c: In function 'entersafe_sanity_check':
../../pkcs15init/pkcs15-entersafe.c:497:33: error: 'profile_auth.attrs.pin.flags' may be used uninitialized [-Werror=maybe-uninitialized]
  497 |                                 && pin_attrs->flags != profile_auth.attrs.pin.flags)   {
      |                                 ^
../../pkcs15init/pkcs15-entersafe.c:475:36: note: 'profile_auth.attrs.pin.flags' was declared here
  475 |         struct sc_pkcs15_auth_info profile_auth;
      |                                    ^
../../pkcs15init/pkcs15-entersafe.c:496:20: error: 'profile_auth.attrs.pin.reference' may be used uninitialized [-Werror=maybe-uninitialized]
  496 |                 if (pin_attrs->reference == profile_auth.attrs.pin.reference
      |                    ^
../../pkcs15init/pkcs15-entersafe.c:475:36: note: 'profile_auth.attrs.pin.reference' was declared here
  475 |         struct sc_pkcs15_auth_info profile_auth;
      |                                    ^
../../pkcs15init/pkcs15-epass2003.c: In function 'epass2003_pkcs15_sanity_check':
../../pkcs15init/pkcs15-epass2003.c:756:21: error: 'profile_auth.attrs.pin.flags' may be used uninitialized [-Werror=maybe-uninitialized]
  756 |                     && pin_attrs->flags != profile_auth.attrs.pin.flags) {
      |                     ^
../../pkcs15init/pkcs15-epass2003.c:732:36: note: 'profile_auth.attrs.pin.flags' was declared here
  732 |         struct sc_pkcs15_auth_info profile_auth;
      |                                    ^
../../pkcs15init/pkcs15-epass2003.c:755:20: error: 'profile_auth.attrs.pin.reference' may be used uninitialized [-Werror=maybe-uninitialized]
  755 |                 if (pin_attrs->reference == profile_auth.attrs.pin.reference
      |                    ^
../../pkcs15init/pkcs15-epass2003.c:732:36: note: 'profile_auth.attrs.pin.reference' was declared here
  732 |         struct sc_pkcs15_auth_info profile_auth;
      |                                    ^
In function 'have_onepin',
    inlined from 'asepcos_create_pin' at ../../pkcs15init/pkcs15-asepcos.c:419:7:
../../pkcs15init/pkcs15-asepcos.c:325:37: error: 'sopin.attrs.pin.flags' may be used uninitialized [-Werror=maybe-uninitialized]
  325 |         if (!(sopin.attrs.pin.flags & SC_PKCS15_PIN_FLAG_SO_PIN))
      |                                     ^
../../pkcs15init/pkcs15-asepcos.c: In function 'asepcos_create_pin':
../../pkcs15init/pkcs15-asepcos.c:321:31: note: 'sopin.attrs.pin.flags' was declared here
  321 |         sc_pkcs15_auth_info_t sopin;
      |                               ^
../../pkcs15init/pkcs15-asepcos.c: In function 'have_onepin':
../../pkcs15init/pkcs15-asepcos.c:325:37: error: 'sopin.attrs.pin.flags' may be used uninitialized [-Werror=maybe-uninitialized]
  325 |         if (!(sopin.attrs.pin.flags & SC_PKCS15_PIN_FLAG_SO_PIN))
      |                                     ^
../../pkcs15init/pkcs15-asepcos.c:321:31: note: 'sopin.attrs.pin.flags' was declared here
  321 |         sc_pkcs15_auth_info_t sopin;
      |                               ^
lto1: all warnings being treated as errors
make[6]: *** [/tmp/ccRMbfeN.mk:71: /tmp/ccxVJcQw.ltrans23.ltrans.o] Error 1
make[6]: *** Waiting for unfinished jobs....

```
this is the error with that variable non memset to 0

Reply:
Thank you for the logs! I think the memset here is more like patch hiding the other problem that the callers of `sc_profile_get_pin_info()` do not initialize the `info` argument and do not check the return values of this function, causing uninitialized reads.

My preference would be either or both
 * Initialize the `info` argument in all calls to this function
 * Check return values of this function before using its results.